regcore_write/views/layer.py

```diff
@@ -21,36 +24,39 @@ def child_label_of(lhs, rhs):
 
 @secure_write
 @json_body
-def add(request, name, label_id, version=None):
+def add(request, name, doc_type, doc_id):
     """Add the layer node and all of its children to the db"""
     layer = request.json_body
-
     if not isinstance(layer, dict):
         return user_error('invalid format')
 
+    params = LayerParams(doc_type, doc_id)
     for key in layer.keys():
         # terms layer has a special attribute
-        if not child_label_of(key, label_id) and key != 'referenced':
-            return user_error('label mismatch: %s, %s' % (label_id, key))
-
-    prefix = label_id
-    if version is not None:
-        prefix = version + ':' + prefix
-    storage.for_layers.bulk_put(
-        child_layers(name, label_id, version, layer), name, prefix)
+        if not child_label_of(key, params.tree_id) and key != 'referenced':
+            return user_error('label mismatch: {}, {}'.format(
+                params.tree_id, key))
 
+    storage.for_layers.bulk_put(child_layers(params, layer), name,
+                                params.doc_type, params.doc_id)
     return success()
 
 
-def child_layers(layer_name, root_label, version, root_layer):
+def child_layers(layer_params, layer_data):
     """We are generally given a layer corresponding to an entire regulation.
     We need to split that layer up and store it per node within the
     regulation. If a reg has 100 nodes, but the layer only has 3 entries, it
     will still store 100 layer models -- many may be empty"""
-    root = storage.for_regulations.get(root_label, version)
-    if not root:
-        root = storage.for_preambles.get(root_label)
-    if not root:
+    doc_id_components = layer_params.doc_id.split('/')
+    if layer_params.doc_type == 'preamble':
+        doc_tree = storage.for_preambles.get(layer_params.doc_id)
+    elif layer_params.doc_type == 'cfr':
+        version, label = doc_id_components
+        doc_tree = storage.for_regulations.get(label, version)
+    else:
+        logging.warning("Unknown doc_type: %s", layer_params.doc_type)
```

Comment:
By convention, prefer named loggers. Also, is it preferable to warn instead of throwing an exception? If this application is running remotely and a user sends up data with invalid document types, they're not going to see this warning (right?).